File: android/app/src/main/kotlin/com/sotalog/android/data/repositories/CallsignHistoryRepository.kt
package com.sotalog.android.data.repositories

import com.sotalog.android.data.local.database.dao.CallsignHistoryDao
import com.sotalog.android.domain.models.CallsignHistory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.Date
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CallsignHistoryRepository @Inject constructor(
    private val callsignHistoryDao: CallsignHistoryDao,
) {

    suspend fun fetch(callsign: String): CallsignHistory? = withContext(Dispatchers.IO) {
        callsignHistoryDao.getByCallsign(callsign.uppercase())
    }

    /**
     * Records that a QSO was just logged: refreshes lastWorked and fills in
     * name/qth/grid if provided. Call this when creating a new QSO.
     */
    suspend fun recordQSO(
        callsign: String,
        name: String?,
        qth: String?,
        grid: String?,
    ) = withContext(Dispatchers.IO) {
        val upper = callsign.uppercase()
        val existing = callsignHistoryDao.getByCallsign(upper)
        if (existing != null) {
            callsignHistoryDao.upsert(
                existing.copy(
                    lastWorked = Date(),
                    name = name?.takeIf { it.isNotEmpty() } ?: existing.name,
                    qth = qth?.takeIf { it.isNotEmpty() } ?: existing.qth,
                    grid = grid?.takeIf { it.isNotEmpty() } ?: existing.grid,
                )
            )
        } else {
            callsignHistoryDao.upsert(
                CallsignHistory(
                    callsign = upper,
                    name = name,
                    qth = qth,
                    grid = grid,
                    lastWorked = Date(),
                )
            )
        }
    }

    /**
     * Updates enrichment from a QRZ lookup or an edit, without touching lastWorked
     * (no new contact was made).
     */
    suspend fun updateFromLookup(
        callsign: String,
        name: String?,
        qth: String?,
        grid: String?,
    ) = withContext(Dispatchers.IO) {
        val upper = callsign.uppercase()
        val existing = callsignHistoryDao.getByCallsign(upper)
        if (existing != null) {
            callsignHistoryDao.upsert(
                existing.copy(
                    name = name?.takeIf { it.isNotEmpty() } ?: existing.name,
                    qth = qth?.takeIf { it.isNotEmpty() } ?: existing.qth,
                    grid = grid?.takeIf { it.isNotEmpty() } ?: existing.grid,
                )
            )
        } else {
            callsignHistoryDao.upsert(
                CallsignHistory(
                    callsign = upper,
                    name = name,
                    qth = qth,
                    grid = grid,
                    lastWorked = null,
                )
            )
        }
    }
}

File: ios/SOTALog/Services/Database/CallsignHistoryRepository.swift
import Foundation
import GRDB

/// Stores cached enrichment (name/QTH/grid/lastWorked) per callsign. The worked
/// count is derived from the `qso` table on demand and is intentionally not kept
/// here — see `QSORepository.countForCallsign`.
struct CallsignHistoryRepository {
    let database: AppDatabase

    /// Fetches the history for a callsign
    func fetch(callsign: String) async throws -> CallsignHistory? {
        try await database.dbWriter.read { db in
            try CallsignHistory.fetchOne(db, id: callsign.uppercased())
        }
    }

    /// Records that a QSO was just logged: refreshes lastWorked and fills in
    /// name/qth/grid if provided. Call this when creating a new QSO.
    func recordQSO(callsign: String, name: String?, qth: String?, grid: String?) async throws {
        try await database.dbWriter.write { db in
            let key = callsign.uppercased()
            if var existing = try CallsignHistory.fetchOne(db, id: key) {
                existing.lastWorked = Date()
                if let name = name, !name.isEmpty { existing.name = name }
                if let qth = qth, !qth.isEmpty { existing.qth = qth }
                if let grid = grid, !grid.isEmpty { existing.grid = grid }
                try existing.update(db)
            } else {
                let history = CallsignHistory(
                    callsign: key,
                    name: name,
                    qth: qth,
                    grid: grid,
                    lastWorked: Date()
                )
                try history.insert(db)
            }
        }
    }

    /// Updates enrichment with data from a QRZ lookup or an edit, without touching
    /// lastWorked (no new contact was made).
    func updateFromLookup(callsign: String, name: String?, qth: String?, grid: String?) async throws {
        try await database.dbWriter.write { db in
            let key = callsign.uppercased()
            if var existing = try CallsignHistory.fetchOne(db, id: key) {
                if let name = name, !name.isEmpty { existing.name = name }
                if let qth = qth, !qth.isEmpty { existing.qth = qth }
                if let grid = grid, !grid.isEmpty { existing.grid = grid }
                try existing.update(db)
            } else {
                let history = CallsignHistory(
                    callsign: key,
                    name: name,
                    qth: qth,
                    grid: grid
                )
                try history.insert(db)
            }
        }
    }
}
